scripts/retrieve_databundle_light.py

```diff
@@ -618,6 +700,35 @@ def datafiles_retrivedatabundle(config):
     return listoutputs
 
 
+def merge_hydrobasins_shape(config):
+    basins_path = config_bundles["bundle_hydrobasins"]["destination"]
+    hydrobasins_level = snakemake.config["renewable"]["hydro"]["resource"][
+        "hydrobasins_level"
+    ]
+
+    mask_file = os.path.join(
+        basins_path, "hybas_*_lev{:02d}_v1c.shp".format(int(hydrobasins_level))
+    )
+    files_to_merge = glob.glob(mask_file)
+
+    gpdf_list = [None] * len(files_to_merge)
+    logger.info("Reading hydrobasins files \n\r")
+    for i in tqdm(range(0, len(files_to_merge))):
+        gpdf_list[i] = gpd.read_file(files_to_merge[i])
+    fl_merged = gpd.GeoDataFrame(pd.concat(gpdf_list))
+    logger.info(
+        "Merging single files into:\n\t"
+        + "hybas_world_lev"
+        + str(hydrobasins_level)
+        + "_v1c.shp"
+    )
+    fl_merged.to_file(
+        os.path.join(
+            basins_path, "hybas_world_lev{:02d}_v1c.shp".format(int(hydrobasins_level))
+        )
```

Comment:
This path should be taken from the config

Reply:
> My suggestion was to propose calling
> download_and_unzip_direct(..., ..., hot_run=hot_run, disable_progress=disable_progress)
> within this for loop, rather than copying the code again, but we can go for the duplication if it is easier

Thanks for the explanation! Have investigated a bit, and it looks like there may be a way to avoid duplication, but we'd need for that to encapsulate a part of `download_and_unzip_direct`. The point is that currently all the downloading functions use the same arguments list `config, rootpath, hot_run=True, disable_progress=False`, which makes possible to use a wildcard when calling the download function:

https://github.com/pypsa-meets-earth/pypsa-earth/blob/5f77c9814d942250371fb8959869370818a8aa48/scripts/retrieve_databundle_light.py#L676-L680

`config` is common for all `download_and_unzip_*` functions, while processing of the config parameters differs for different functions. So, I have added `download_and_unpack` function to replace duplicated functionality both in `download_and_unzip_direct` and `download_and_unzip_hydrobasins`. Would you agree with this approach?

Reply:
> This path should be taken from the config

An understanding question: do you mean using `snakemake.config["renewable"]["hydro"]["resource"]["hydrobasins"]`? 

Agree that it's the major parameter transferred to atlite. However, there is also `hydrobasins_level` parameter which basically duplicate `hydrobasins`. So, it's perfectly possible to organise in the `config.yaml` a data mismatch: like set `hydrobasins_level: 4` along with `hydrobasins: data/hydrobasins/hybas_world_lev06_v1c.shp`. If we'll take the file name from the config, this error will remain unnoticed, while if an actual  `hydrobasins_level` is used to generate the merged file, we keep the prepared data consistent, while there will be an error thrown while building renewable profiles. To me, this design look more error-safe.

What is your feeling about that? (Sorry for bringing the complexity which is probably not needed there 🙃)

Reply:
I quite believe that in our case we should have a file like: data/hydrobasins/hybas_world.shp that is independent from the level.
It is the merge of all hydrobasin shapes regardless of the level.
If the level is modified, the function is triggered again, that should simplify the procedure